pydra/engine/core.py

```diff
@@ -797,6 +797,66 @@ def _reset(self):
                 task._reset()
 
 
+def _sanitize_input_spec(
+    input_spec: ty.Union[BaseSpec, ty.List[str]],
+    wf_name: str,
+) -> BaseSpec:
+    """Makes sure the provided input specifications are valid.
+
+    If the input specification is a list of strings, this will
+    build a proper SpecInfo object out of it.
+
+    Parameters
+    ----------
+    input_spec : BaseSpec or List[str]
+        Input specification to be sanitized.
+
+    wf_name : str
+        The name of the workflow for which the input specifications
+        are sanitized.
+
+    Returns
+    -------
+    input_spec : BaseSpec
+        Sanitized input specifications.
+
+    Raises
+    ------
+    ValueError
+        If provided `input_spec` is None.
+    """
+    graph_checksum_input = ("_graph_checksums", ty.Any)
+    if input_spec:
+        if isinstance(input_spec, BaseSpec):
```

Comment:
It looks like this line wasn't tested. I tried adding a new test for it, but it seems like giving a `BaseSpec` or a derived class instance to the `Workflow` constructor gives an error:


```python
>>> from pydra.engine.specs import ShellSpec
>>> spec = ShellSpec(executable="foo")
>>> from pydra.engine import Workflow
>>> wf = Workflow(name="bar", input_spec=spec)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/core.py", line 917, in __init__
    super().__init__(
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/core.py", line 140, in __init__
    klass = make_klass(self.input_spec)
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/helpers.py", line 258, in make_klass
    fields = spec.fields
AttributeError: 'ShellSpec' object has no attribute 'fields'
```

Reply:
the previous PR change that was merged specifically insisted on `BaseSpec`, since a Workflow should never require anything else. so a `ShellSpec` instance in theory should not be passed to a Workflow. do you get the same issue if you use a `BaseSpec`? (i don't think a release was made since that change, so you would have to test with the github master branch)

Reply:
Yes, using a `BaseSpec` or any derived class instance gives the same error:

```python
>>> from pydra.engine.specs import BaseSpec
>>> from pydra.engine import Workflow
>>> wf = Workflow(name="bar", input_spec=BaseSpec())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/core.py", line 883, in __init__
    super().__init__(
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/core.py", line 140, in __init__
    klass = make_klass(self.input_spec)
  File "/Users/nicolas.gensollen/GitRepos/pydra/pydra/engine/helpers.py", line 258, in make_klass
    fields = spec.fields
AttributeError: 'BaseSpec' object has no attribute 'fields'
```

I am on the latest commit on master.

Reply:
sorry, this indeed needs some update. it cannot be a `BaseSpec` by itself, it needs to be something of this nature: https://github.com/nipype/pydra/blob/master/pydra/engine/core.py#L855 

In fact the `SpecInfo` version should also check for graph_checksums as a field. i think this is primarily why we avoided allowing this to simply be a named list, but if you are up for it to fix this up in your PR that would be great.

technically any input needs to be a SpecInfo, hence BaseSpec should probably be removed from the typing list as well.

i know @tclose looked at this recently, so please chime in.

Reply:
Thanks for your fast reply @satra. If I understand correctly, the user shouldn't be able to pass a `BaseSpec` instance to the `Workflow` constructor, but either a `SpecInfo` instance, or a list of strings which is then converted to a proper `SpecInfo` instance. 

So, unless I missed something, this is what needs to be done:

- Remove `BaseSpec` from the type hint and update the docstring accordingly
- Raise a `ValueError` when a `BaseSpec` instance is passed to the constructor (basically remove lines 830 and 831 in this PR)
- Make sure the input specifications have a `_graph_checksums` attribute (implemented in this PR for `SpecInfo` and already implemented for list of strings).

Reply:
thank you @NicolasGensollen  for tracking this down.  your steps are right. i don't think it makes sense to have `BaseSpec` as an input, but it's been a long time since we wrote that so the thinking process is no longer in memory.